<div align='center'><picture><h1><b>Data Analysis Process</b></h1></picture></div>

# **Article 129 : Data Gathering** [![Static Badge](https://img.shields.io/badge/Open%20in%20Colab%20-%20orange?style=plastic&logo=googlecolab&labelColor=grey)](https://colab.research.google.com/github/sshrizvi/DataScienceMastery/blob/main/DataAnalysisProcess/Notebooks/129_data_gathering.ipynb)

|🔴 **NOTE** 🔴|
|:-----------:|
| This notebook contains the practical implementations of the concepts discussed in the following article.|
| Here is Article 129 - [Data Gathering](../Articles/129_data_gathering.md) |

### 📦 **Importing Relevant Libraries**

In [2]:
import pandas as pd

## 🧶 **Importing Data**

In the Data Analysis Process (DAP), we usually deal with data in the form of CSVs. We just learnt the way to read CSV files into DataFrames with the help of `pd.read_csv()` method. Data are available in many formats and, importing them is a little bit different from one another.

In this notebook, we will discuss various ways to import data of various formats, from various sources.

### 🎯 **Importing CSVs**

The first question that comes in my mind is that, what is this CSV?  
And the answer to this question can be given with the help of this explanation.  

CSV stands for, *Comma Separated Values*. It is a type of file used to store tabular data.  
In CSVs, the data of a row is stored in a single line where values are separated by commas. Every row is separated with one another, using the new line.  

| ID      | Age   | Gender |
| ------- | ---   | ------ |
| 1       | 23    | M      |
| 2       | 24    | F      |
| 3       | 25    | M      |

Here is a sample of CSV file for the above tabular data:  
```csv
ID, Age, Gender
1, 23, M
2, 24, F
3, 25, M
```

Now, lets learn in detail about the `pandas.read_csv()` method, that is used to load a CSV file as Pandas DataFrame.

#### **1. Importing a Local CSV File**

To load a local CSV file using `pd.read_csv` method, we will have to provide the path of the file to the `filepath_or_buffer` parameter. Lets see this.

In [3]:
nifty_df = pd.read_csv(
    filepath_or_buffer='../Resources/Data/NIFTY50.csv'
)

The `filepath_or_buffer` parameter is the first argument position wise. So, usually we do not mention the parameter name while passing the path to the CSV file.

In [4]:
nifty_df = pd.read_csv('../Resources/Data/NIFTY50.csv')

#### **2. Importing a Remote CSV File**

Loading a remote CSV file is way easier than we think of it. All we have to do is to just pass the remote URL to the `filepath_or_buffer` parameter. `pd.read_csv()` is much flexible and is designed in the way that it can load a remote CSV file, as if we are loading a local file.

In [5]:
countries_df = pd.read_csv('https://raw.githubusercontent.com/cs109/2014_data/master/countries.csv')

#### **3. The `sep` parameter**

As we know that the separator, in case of a *CSV File*, is comma (,). But, more often we have to deal with file formats like TSVs (Tab Separated Values) and more formats containing different separators.  
So, for loading those files using `pd.read_csv()` becomes easier with the `sep` or `delimeter` parameter.  
Here is a demonstration for loading a sample TSV file.

In [6]:
tsv_df = pd.read_csv('../Resources/Data/file.tsv', sep='\t',
                     names=['column_a', 'column_b', 'column_c', 'column_d'])

Note that the `names` parameter used here is used to provide column names explicitly. We used this because in `files.tsv`, we do not have the column names.

#### **4. The `index_col` parameter**

This parameter of `pd.read_csv()` parameter allows us to use any of the columns of the dataset to use as row label(s), or index.  
Here is a demonstration to it.

In [7]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv')

Here, you can see that a default `Index` Object is created for row labels, when we did not pass the `index_col` parameter, a column name.

In [8]:
train_df.head()

enrollee_id      city  city_development_index gender  \
0         8949  city_103                   0.920   Male   
1        29725   city_40                   0.776   Male   
2        11561   city_21                   0.624    NaN   
3        33241  city_115                   0.789    NaN   
4          666  city_162                   0.767   Male   

       relevent_experience enrolled_university education_level  \
0  Has relevent experience       no_enrollment        Graduate   
1   No relevent experience       no_enrollment        Graduate   
2   No relevent experience    Full time course        Graduate   
3   No relevent experience                 NaN        Graduate   
4  Has relevent experience       no_enrollment         Masters   

  major_discipline experience company_size    company_type last_new_job  \
0             STEM        >20          NaN             NaN            1   
1             STEM         15        50-99         Pvt Ltd           >4   
2             STEM          5          NaN             NaN        never   
3  Business Degree         <1          NaN         Pvt Ltd        never   
4             STEM        >20        50-99  Funded Startup            4   

   training_hours  target  
0              36     1.0  
1              47     0.0  
2              83     0.0  
3              52     1.0  
4               8     0.0

Now, lets pass the `enrollee_id` to `index_col` to make it the Index.

In [9]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv', index_col='enrollee_id')

In [10]:
train_df.head()

city  city_development_index gender      relevent_experience  \
enrollee_id                                                                     
8949         city_103                   0.920   Male  Has relevent experience   
29725         city_40                   0.776   Male   No relevent experience   
11561         city_21                   0.624    NaN   No relevent experience   
33241        city_115                   0.789    NaN   No relevent experience   
666          city_162                   0.767   Male  Has relevent experience   

            enrolled_university education_level major_discipline experience  \
enrollee_id                                                                   
8949              no_enrollment        Graduate             STEM        >20   
29725             no_enrollment        Graduate             STEM         15   
11561          Full time course        Graduate             STEM          5   
33241                       NaN        Graduate  Business Degree         <1   
666               no_enrollment         Masters             STEM        >20   

            company_size    company_type last_new_job  training_hours  target  
enrollee_id                                                                    
8949                 NaN             NaN            1              36     1.0  
29725              50-99         Pvt Ltd           >4              47     0.0  
11561                NaN             NaN        never              83     0.0  
33241                NaN         Pvt Ltd        never              52     1.0  
666                50-99  Funded Startup            4               8     0.0

#### **5. The `usecols` parameter**

The `usecols` parameter helps us to import only selective columns of our choice from the dataset.

In [11]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv',
            usecols=['enrollee_id', 'gender', 'education_level'])

#### **6. The `header` parameter**

The `header` parameter takes an integer or a sequence of integers as argument, that represents the row number *(or index)* from where the start of the data is assumed. It is useful when the first row of the CSV file represents tha column names and is not actually a part of the rows of the dataset.

The `header` parameter takes the following arguments :  
1. `int` : A scalar integer value, pointing towards the start of the dataset.
2. `sequence of ints` : A sequence of integer value is provided in case of `MultiIndex`.
3. `infer` : The default one, automatically parses the first row of the CSV file as the column header, and parses its values as column lables.
4. `None` : It is automatically used if the `names` parameter is given. Explicitly assigning, `header=None`, without the `names` parameter will lead to an Index Object filled with integer values.

In [12]:
test_df = pd.read_csv('../Resources/Data/test.csv')

In [13]:
test_df

Unnamed: 0   Unnamed: 1 Unnamed: 2              Unnamed: 3 Unnamed: 4  \
0           0  enrollee_id       city  city_development_index     gender   
1           1        29725    city_40                   0.776       Male   
2           2        11561    city_21                   0.624        NaN   
3           3        33241   city_115                   0.789        NaN   
4           4          666   city_162                   0.767       Male   

                Unnamed: 5           Unnamed: 6       Unnamed: 7  \
0      relevent_experience  enrolled_university  education_level   
1   No relevent experience        no_enrollment         Graduate   
2   No relevent experience     Full time course         Graduate   
3   No relevent experience                  NaN         Graduate   
4  Has relevent experience        no_enrollment          Masters   

         Unnamed: 8  Unnamed: 9   Unnamed: 10     Unnamed: 11   Unnamed: 12  \
0  major_discipline  experience  company_size    company_type  last_new_job   
1              STEM          15         50-99         Pvt Ltd            >4   
2              STEM           5           NaN             NaN         never   
3   Business Degree          <1           NaN         Pvt Ltd         never   
4              STEM         >20         50-99  Funded Startup             4   

      Unnamed: 13 Unnamed: 14  
0  training_hours      target  
1              47           0  
2              83           0  
3              52           1  
4               8           0

Now, with the `header` parameter.

In [14]:
test_df = pd.read_csv('../Resources/Data/test.csv', header=1, index_col='0')

In [15]:
test_df

enrollee_id      city  city_development_index gender  \
0                                                         
1        29725   city_40                   0.776   Male   
2        11561   city_21                   0.624    NaN   
3        33241  city_115                   0.789    NaN   
4          666  city_162                   0.767   Male   

       relevent_experience enrolled_university education_level  \
0                                                                
1   No relevent experience       no_enrollment        Graduate   
2   No relevent experience    Full time course        Graduate   
3   No relevent experience                 NaN        Graduate   
4  Has relevent experience       no_enrollment         Masters   

  major_discipline experience company_size    company_type last_new_job  \
0                                                                         
1             STEM         15        50-99         Pvt Ltd           >4   
2             STEM          5          NaN             NaN        never   
3  Business Degree         <1          NaN         Pvt Ltd        never   
4             STEM        >20        50-99  Funded Startup            4   

   training_hours  target  
0                          
1              47       0  
2              83       0  
3              52       1  
4               8       0

#### **7. The `skiprows` parameter**

It is an *optional* parameter of the `pd.read_csv()` method, that allows us to skip rows from getting loaded into the DataFrame.  

Here are the values that it can take as an argument :  
1. `int` : Providing a scalar integer, allows you to skip that much starting rows of the dataset.
2. `sequence of ints` : A sequence of integers is provided only to specify particular rows to be skipped.
3. `callable` : A function returning `True` or `False`. Rows for which the function returns `True` will be skipped.

In [16]:
skipped_df = pd.read_csv('../Resources/Data/aug_train.csv', skiprows=10)
skipped_df

699  city_103   0.92 Unnamed: 3  Has relevent experience  \
0      29452   city_21  0.624        NaN   No relevent experience   
1      23853  city_103  0.920       Male  Has relevent experience   
2      25619   city_61  0.913       Male  Has relevent experience   
3       5826   city_21  0.624       Male   No relevent experience   
4       8722   city_21  0.624        NaN   No relevent experience   
...      ...       ...    ...        ...                      ...   
19143   7386  city_173  0.878       Male   No relevent experience   
19144  31398  city_103  0.920       Male  Has relevent experience   
19145  24576  city_103  0.920       Male  Has relevent experience   
19146   5756   city_65  0.802       Male  Has relevent experience   
19147  23834   city_67  0.855        NaN   No relevent experience   

          no_enrollment        Graduate        STEM   17     10000+  Pvt Ltd  \
0      Full time course     High School         NaN    2        NaN      NaN   
1         no_enrollment        Graduate        STEM    5  5000-9999  Pvt Ltd   
2         no_enrollment        Graduate        STEM  >20  1000-4999  Pvt Ltd   
3                   NaN             NaN         NaN    2        NaN      NaN   
4      Full time course     High School         NaN    5        NaN      NaN   
...                 ...             ...         ...  ...        ...      ...   
19143     no_enrollment        Graduate  Humanities   14        NaN      NaN   
19144     no_enrollment        Graduate        STEM   14        NaN      NaN   
19145     no_enrollment        Graduate        STEM  >20      50-99  Pvt Ltd   
19146     no_enrollment     High School         NaN   <1    500-999  Pvt Ltd   
19147     no_enrollment  Primary School         NaN    2        NaN      NaN   

          >4  123  0.0  
0      never   32  1.0  
1          1  108  0.0  
2          3   23  0.0  
3      never   24  0.0  
4      never   26  0.0  
...      ...  ...  ...  
19143      1   42  1.0  
19144      4   52  1.0  
19145      4   44  0.0  
19146      2   97  0.0  
19147      1  127  0.0  

[19148 rows x 14 columns]

In [17]:
skipped_df = pd.read_csv('../Resources/Data/aug_train.csv', skiprows=[0, 1, 2, 3, 4])
skipped_df

666  city_162  0.767  Male  Has relevent experience  \
0      21651  city_176  0.764   NaN  Has relevent experience   
1      28806  city_160  0.920  Male  Has relevent experience   
2        402   city_46  0.762  Male  Has relevent experience   
3      27107  city_103  0.920  Male  Has relevent experience   
4        699  city_103  0.920   NaN  Has relevent experience   
...      ...       ...    ...   ...                      ...   
19148   7386  city_173  0.878  Male   No relevent experience   
19149  31398  city_103  0.920  Male  Has relevent experience   
19150  24576  city_103  0.920  Male  Has relevent experience   
19151   5756   city_65  0.802  Male  Has relevent experience   
19152  23834   city_67  0.855   NaN   No relevent experience   

          no_enrollment         Masters        STEM  >20    50-99  \
0      Part time course        Graduate        STEM   11      NaN   
1         no_enrollment     High School         NaN    5    50-99   
2         no_enrollment        Graduate        STEM   13      <10   
3         no_enrollment        Graduate        STEM    7    50-99   
4         no_enrollment        Graduate        STEM   17   10000+   
...                 ...             ...         ...  ...      ...   
19148     no_enrollment        Graduate  Humanities   14      NaN   
19149     no_enrollment        Graduate        STEM   14      NaN   
19150     no_enrollment        Graduate        STEM  >20    50-99   
19151     no_enrollment     High School         NaN   <1  500-999   
19152     no_enrollment  Primary School         NaN    2      NaN   

       Funded Startup   4    8  0.0  
0                 NaN   1   24  1.0  
1      Funded Startup   1   24  0.0  
2             Pvt Ltd  >4   18  1.0  
3             Pvt Ltd   1   46  1.0  
4             Pvt Ltd  >4  123  0.0  
...               ...  ..  ...  ...  
19148             NaN   1   42  1.0  
19149             NaN   4   52  1.0  
19150         Pvt Ltd   4   44  0.0  
19151         Pvt Ltd   2   97  0.0  
19152             NaN   1  127  0.0  

[19153 rows x 14 columns]

In [18]:
skipped_df = pd.read_csv('../Resources/Data/aug_train.csv', skiprows=lambda x : True if x % 2 == 0 else False)
skipped_df

8949  city_103   0.92    Male  Has relevent experience  \
0     11561   city_21  0.624     NaN   No relevent experience   
1       666  city_162  0.767    Male  Has relevent experience   
2     28806  city_160  0.920    Male  Has relevent experience   
3     27107  city_103  0.920    Male  Has relevent experience   
4     29452   city_21  0.624     NaN   No relevent experience   
...     ...       ...    ...     ...                      ...   
9573   9212   city_21  0.624     NaN  Has relevent experience   
9574  32313  city_160  0.920  Female  Has relevent experience   
9575  29754  city_103  0.920  Female  Has relevent experience   
9576  31398  city_103  0.920    Male  Has relevent experience   
9577   5756   city_65  0.802    Male  Has relevent experience   

         no_enrollment     Graduate        STEM  >20 Unnamed: 9  \
0     Full time course     Graduate        STEM    5        NaN   
1        no_enrollment      Masters        STEM  >20      50-99   
2        no_enrollment  High School         NaN    5      50-99   
3        no_enrollment     Graduate        STEM    7      50-99   
4     Full time course  High School         NaN    2        NaN   
...                ...          ...         ...  ...        ...   
9573     no_enrollment      Masters        STEM    3    100-500   
9574     no_enrollment     Graduate        STEM   10    100-500   
9575     no_enrollment     Graduate  Humanities    7      10/49   
9576     no_enrollment     Graduate        STEM   14        NaN   
9577     no_enrollment  High School         NaN   <1    500-999   

         Unnamed: 10      1  36  1.0  
0                NaN  never  83  0.0  
1     Funded Startup      4   8  0.0  
2     Funded Startup      1  24  0.0  
3            Pvt Ltd      1  46  1.0  
4                NaN  never  32  1.0  
...              ...    ...  ..  ...  
9573         Pvt Ltd      3  40  1.0  
9574   Public Sector      3  23  0.0  
9575  Funded Startup      1  25  0.0  
9576             NaN      4  52  1.0  
9577         Pvt Ltd      2  97  0.0  

[9578 rows x 14 columns]

#### **8. The `nrows` parameter**

This optional parameter takes an `integer` value, representing the number of lines to read from the starting.

In [19]:
nrows_df = pd.read_csv('../Resources/Data/aug_train.csv', nrows=5)
nrows_df.shape

(5, 14)

#### **9. The `encoding` parameter**

This parameter helps us to provide explicit *encoding schemes* like `UTF-8`.

In [20]:
zomato_df = pd.read_csv('../Resources/Data/zomato.csv', encoding='latin-1')
zomato_df

Restaurant ID           Restaurant Name  Country Code              City  \
0           6317637          Le Petit Souffle           162       Makati City   
1           6304287          Izakaya Kikufuji           162       Makati City   
2           6300002    Heat - Edsa Shangri-La           162  Mandaluyong City   
3           6318506                      Ooma           162  Mandaluyong City   
4           6314302               Sambo Kojin           162  Mandaluyong City   
...             ...                       ...           ...               ...   
9546        5915730              NamlÛ± Gurme           208         ÛÁstanbul   
9547        5908749             Ceviz AÛôacÛ±           208         ÛÁstanbul   
9548        5915807                     Huqqa           208         ÛÁstanbul   
9549        5916112              Aôôk Kahve           208         ÛÁstanbul   
9550        5927402  Walter's Coffee Roastery           208         ÛÁstanbul   

                                                Address  \
0     Third Floor, Century City Mall, Kalayaan Avenu...   
1     Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2     Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3     Third Floor, Mega Fashion Hall, SM Megamall, O...   
4     Third Floor, Mega Atrium, SM Megamall, Ortigas...   
...                                                 ...   
9546  Kemankeô Karamustafa Paôa Mahallesi, RÛ±htÛ±...   
9547  Koôuyolu Mahallesi, Muhittin íìstí_ndaÛô Cadd...   
9548  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9549  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9550  CafeaÛôa Mahallesi, BademaltÛ± Sokak, No 21/B,...   

                                        Locality  \
0      Century City Mall, Poblacion, Makati City   
1     Little Tokyo, Legaspi Village, Makati City   
2     Edsa Shangri-La, Ortigas, Mandaluyong City   
3         SM Megamall, Ortigas, Mandaluyong City   
4         SM Megamall, Ortigas, Mandaluyong City   
...                                          ...   
9546                                    Karakí_y   
9547                                   Koôuyolu   
9548                                 Kuruí_eôme   
9549                                 Kuruí_eôme   
9550                                        Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall, Poblacion, Makati City, Mak...  121.027535   
1     Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101   
2     Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831   
3     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475   
4     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508   
...                                                 ...         ...   
9546                                Karakí_y, ÛÁstanbul   28.977392   
9547                               Koôuyolu, ÛÁstanbul   29.041297   
9548                             Kuruí_eôme, ÛÁstanbul   29.034640   
9549                             Kuruí_eôme, ÛÁstanbul   29.036019   
9550                                    Moda, ÛÁstanbul   29.026016   

       Latitude                          Cuisines  ...          Currency  \
0     14.565443        French, Japanese, Desserts  ...  Botswana Pula(P)   
1     14.553708                          Japanese  ...  Botswana Pula(P)   
2     14.581404  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)   
3     14.585318                   Japanese, Sushi  ...  Botswana Pula(P)   
4     14.584450                  Japanese, Korean  ...  Botswana Pula(P)   
...         ...                               ...  ...               ...   
9546  41.022793                           Turkish  ...  Turkish Lira(TL)   
9547  41.009847   World Cuisine, Patisserie, Cafe  ...  Turkish Lira(TL)   
9548  41.055817            Italian, World Cuisine  ...  Turkish Lira(TL)   
9549  41.057979                   Restaurant Cafe  ...  Turkish Lira(TL)   
9550  40.9

#### **10. The `error_bad_lines` parameter**

This parameter of `pd.read_csv` method helps us to skip bad lines, while importing CSV.  
Specifies what to do upon encountering a bad line (a line with too many fields).  

Allowed values are :
- `error` :  raise an Exception when a bad line is encountered.
- `warn` :  raise a warning when a bad line is encountered and skip that line.
- `skip` :  skip bad lines without raising or warning when they are encountered.

In [21]:
books_df = pd.read_csv('../Resources/Data/bad_lines.csv', on_bad_lines='skip')
books_df

id   age   income
0   1    20    20000
1   2    30    30000
2   3    40    40000
3   5    60    60000

#### **11. The `dtype` parameter**

This *optional* parameter allows us to explicitly apply data types to the columns of the dataset, we are importing.  
**Note :** Keep in mind that if `convertors` are used then `dtype` for those columns will be skipped.

In [22]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv',
                       dtype={'target': int})
train_df['target'].dtype

dtype('int32')

#### **12. The `parse_dates` parameter**

This parameter allows us to parse particular columns of the dataset as dates.

In [23]:
ipl_df = pd.read_csv('../Resources/Data/IPL Matches 2008-2020.csv',
                     parse_dates=['date'])
ipl_df['date'].dtype

dtype('<M8[ns]')

#### **13. The `convertors` parameter**

This *optional* parameter of `pd.read_csv()` method allows us to apply functions to the columns of the dataset while importing. It takes a dictionary of column label as keys and function as values.  
Here is a demonstration of it.

In [24]:
def experience_converter(exp):
    if exp == 'Has relevent experience':
        return 'YES'
    else:
        return 'NO'

In [25]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv',
                       converters=dict(relevent_experience=experience_converter))
train_df['relevent_experience'].unique()

array(['YES', 'NO'], dtype=object)

#### **14. The `na_values` parameter**

It helps us to tell Pandas that you will also have to treat these values as NaNs.  
For an instance, lets consider that `Male` is a `NaN` value. So here is how to tell it to `read_csv()` method while importing.

In [26]:
train_df = pd.read_csv('../Resources/Data/aug_train.csv', na_values='Male')
train_df['gender'].unique()

array([nan, 'Female', 'Other'], dtype=object)

#### **15. The `chunksize` parameter**

While working with CSVs, we often come across CSVs requiring much RAM to get loaded, than of our system.  
So, we use the `chunksize` parameter to load that huge CSV in chunks. Here is how it works.  

**Note :** It returns a `TextFileReader` object for iteration on chunks.

In [27]:
train_chunks = pd.read_csv('../Resources/Data/aug_train.csv', chunksize=5000)
for chunk in train_chunks:
    print(chunk.shape)

(5000, 14)
(5000, 14)
(5000, 14)
(4158, 14)


### 🎯 **Importing XLSX**

As we know that `.xlsx` is the extension used for the Microsoft Excel Files. To import these files using Pandas, we do have a method named `pd.read_excel()`. The interesting thing is that, it works just like `pd.read_csv()` method. Most of the parameters that we saw earlier in this notebook works the same way in this method as in `pd.read_csv` method.

In [28]:
sheet1_df = pd.read_excel('../Resources/Data/students.xlsx')
sheet1_df

Student ID    Student Name  Math Marks  Science Marks  History Marks  \
0       S001     Anya Sharma          92             85             78   
1       S002     Rohan Patel          75             68             89   
2       S003       Zara Khan          98             95             91   
3       S004  Liam O'Connell          55             60             45   
4       S005      Chloe Wong          88             91             82   
5       S006    Ethan Miller          65             72             70   

   Total Marks  Passed All Subjects  
0          255                 True  
1          232                 True  
2          284                 True  
3          160                False  
4          261                 True  
5          207                 True

There is a special parameter named `sheet_name`, that allows us to read data from a particular sheet of the excel file.

In [29]:
sheet2_df = pd.read_excel('../Resources/Data/students.xlsx',
                          sheet_name='Sheet2')
sheet2_df

Student Name  Average Marks Student ID Enrollment Date Course Status  \
0     Anya Sharma          85.00       S001      2024-08-15        Active   
1     Rohan Patel          77.33       S002      2024-09-01        Active   
2       Zara Khan          94.67       S003      2024-07-20        Active   
3  Liam O'Connell          53.33       S004      2024-10-10  On Probation   
4      Chloe Wong          87.00       S005      2024-08-25        Active   
5    Ethan Miller          69.00       S006      2024-09-15        Active   

   Scholarship (%)  
0               50  
1               25  
2               75  
3                0  
4               50  
5               10

### 🎯 **Importing TXT**

More often, because of disorganization, we get to see that tabular data is placed in a Text file with extension `.txt`. So, to read such kind of files, we will inspect the text file for the delimeter (*or separator*). According to it, we will use the parameters of the `read_csv` method to read that text file properly.

In [30]:
tsv_df = pd.read_csv('../Resources/Data/file.txt', sep='\t',
                     names=['column_a', 'column_b', 'column_c', 'column_d'])
tsv_df.head()

column_a  column_b  column_c   column_d
0         0        50         5  881250949
1         0       172         5  881250949
2         0       133         1  881250949
3       196       242         3  881250949
4       186       302         3  891717742

### 🎯 **Importing JSON**

**JSON** stands for **JavaScript Object Notation**.
It is a **lightweight**, **text-based data format** used to **store and exchange data** between systems, especially between a **server and a client**.

JSON is **easy for humans to read and write** and **easy for machines to parse and generate**, which is why it is widely used in web applications, APIs, and data science workflows.  

Most of the data fetched using APIs are returned in the form of JSON. So, we must learn to import JSON files using Pandas. For this purpose Pandas provide a method `pd.read_json()`, with the help of which we can easily import JSON data into Pandas DataFrames.

Here is a demonstration.

In [31]:
json_df = pd.read_json('../Resources/Data/train.json')
json_df.head()

id      cuisine                                        ingredients
0  10259        greek  [romaine lettuce, black olives, grape tomatoes...
1  25693  southern_us  [plain flour, ground pepper, salt, tomatoes, g...
2  20130     filipino  [eggs, pepper, salt, mayonaise, cooking oil, g...
3  22213       indian                [water, vegetable oil, wheat, salt]
4  13162       indian  [black pepper, shallots, cornflour, cayenne pe...

### ⚠️ **Data Warning**  
For the SQL section, we will be using the WORLD Dataset available on Kaggle.  
Here is the link to the dataset : [LINK](https://www.kaggle.com/datasets/busielmorley/worldcities-pop-lang-rank-sql-create-tbls)  
You can also find the SQL file in the [Data](../Resources/Data/) folder.

### 🎯 **Importing SQL**

SQL, stands for *Structured Query Language*, is a type of file that uses SQL files to create tables or databases.  
Importing SQL files is a little bit different from other file types.

Suppose, we have a `.sql` dump file. Unlike other file types, `.sql` files cannot be imported directly using Pandas methods.  
Now the question is, how can we import the `.sql` file into Pandas DataFrame. Here is a solution to this.  

**THE INTUITION :** SQL files can be easily dumped in a SQL database. Also, pandas can easily communicate to SQL database using a connection object.

**THE SOLUTION :** Here are the steps that we need to follow in order to import SQL files into Pandas DataFrame.  
1. Install a SQL database.  
   (*In my case, I have installed MySQL Server 5.7 & MySQL Workbench*)
2. Setup *username* and *password*, while installation.
3. Open *MySQL Workbench*. You will be able to see a default connection. 
 
![alt text](<../Resources/Images/Screenshot 2026-01-09 161610.png>)

1. You will need following things ready with you before moving forward.
   1. Username : (*In my case, its **root***)
   2. Host : (*In my case, its **localhost***)
   3. Port : (*In my case, its **3306***)
   4. Password : (*In my case, its **123456***)
2. Now, connect to a MySQL server using Workbench, or in a way you like.
3. Open the `.sql` file and run it, so that the database and tables are created properly. (*These created database and tables will be fetched further using Pandas and SQLAlchemy*)
4. Now, install `sqlalchemy` using `pip` or `conda`, as per your working environment.
   ```bash
   pip install sqlalchemy
   ```
   OR
   ```bash
   conda install sqlalchemy
   ```
   For more reference, read these : [Official SQLAlchemy DOCS](https://docs.sqlalchemy.org/en/20/), [Connections using SQLAlchemy](https://docs.sqlalchemy.org/en/20/core/connections.html), [Connection String Format for Engine Creation](https://docs.sqlalchemy.org/en/20/dialects/mysql.html#module-sqlalchemy.dialects.mysql.mysqlconnector)
5. Now, install a dialect/DBAPI, as it is required by SQLAlchemy. In my case, I am installing `mysql-connector-python`.
   ```bash
   pip install mysql-connector-python
   ```
   OR
   ```bash
   conda install mysql-connector-python
   ```
6.  Once, installed following things needs to be done.

10. Import `create_engine` method from `sqlalchemy`.

In [32]:
from sqlalchemy import create_engine

11. Create an engine using `create_engine` method of `sqlalchemy`.

In [33]:
engine = create_engine('mysql+mysqlconnector://root:62292003@localhost:3306/world')

12. Once, an engine is created, you can use that engine with Pandas *SQL Reading Methods* like `pd.read_sql()`, `pd.read_sql_query()`, and `pd.read_sql_table()`.

In [34]:
city_df = pd.read_sql(sql='city',
                      con=engine)
city_df.head()

ID            Name CountryCode       District  Population
0   1           Kabul         AFG          Kabol     1780000
1   2        Qandahar         AFG       Qandahar      237500
2   3           Herat         AFG          Herat      186800
3   4  Mazar-e-Sharif         AFG          Balkh      127800
4   5       Amsterdam         NLD  Noord-Holland      731200

Now, lets understand these three SQL Methods of Pandas.  
1. `pd.read_sql_table()` : It is the method that uses the connection/engine object and requires a table name to fetch from the database.
2. `pd.read_sql_query()` : It is the method that uses the connection/engine object and requires a SQL Query to fetch data from the database based on the SQL Query.
3. `pd.read_sql()` : It is the method that uses the connection/engine object and delegates the value of its `sql` parameter to respective two methods listed above.
   1. If the user provides a SQL Table Name, it will delegate the task to `pd.read_sql_table()` method.
   2. If the user provides a SQL Query, it will delegate the task to `pd.read_sql_query()` method.

Here are some demonstration of each.

Fetching the `country` table from the MySQL Database using `pd.read_sql_table()` method.

In [35]:
country_df = pd.read_sql_table(table_name='country',
                               con=engine)
country_df.shape

(239, 15)

Fetching the `country` table from the MySQL Database using `pd.read_sql()` method.

In [36]:
country_df = pd.read_sql(sql='country',
                         con=engine)
country_df.shape

(239, 15)

Fetching the data where Life Expectancy of Countries is greater than 60, using the `pd.read_sql_query()` method.

In [37]:
country_60 = pd.read_sql_query(sql='''
                                SELECT Name, Continent, LifeExpectancy
                                FROM country
                                WHERE LifeExpectancy > 60;
                                ''',
                                con=engine)
country_60.head()

Name      Continent  LifeExpectancy
0                 Aruba  North America            78.4
1              Anguilla  North America            76.1
2               Albania         Europe            71.6
3               Andorra         Europe            83.5
4  Netherlands Antilles  North America            74.7

Fetching the data where Life Expectancy of Countries is greater than 60, using the `pd.read_sql()` method.

In [38]:
country_60 = pd.read_sql(sql='''
                         SELECT Name, Continent, LifeExpectancy
                         FROM country
                         WHERE LifeExpectancy > 60;
                         ''',
                         con=engine)
country_60.head()

Name      Continent  LifeExpectancy
0                 Aruba  North America            78.4
1              Anguilla  North America            76.1
2               Albania         Europe            71.6
3               Andorra         Europe            83.5
4  Netherlands Antilles  North America            74.7

## 🧶 **Exporting Data**

While working with data, we often need to export and save our processed data to some kind of file formats. So here are some formats in which we can export and save our data using Pandas.

### 🎯 **Exporting CSVs**

To export your processed CSV using pandas, you will have to use the `pandas.DataFrame.to_csv()` method.

In [39]:
del_df = pd.read_csv('../Resources/Data/deliveries.csv')

So, lets first prepare a sample CSV from the loaded `deliveries` dataset.  
For this purpose, we are going to create a DataFrame containing total runs scored by each batsman.

In [40]:
batsman_runs_history = del_df.groupby('batsman')['batsman_runs'].sum().reset_index()
batsman_runs_history.head()

batsman  batsman_runs
0  A Ashish Reddy           280
1      A Chandila             4
2        A Chopra            53
3     A Choudhary            25
4     A Dananjaya             4

Now, lets export it to a csv file, so that anyone can use it for analysis.

In [41]:
batsman_runs_history.to_csv('../Resources/Exports/batsman_runs_history.csv')

#### **1. The `index` parameter**

Now, lets verify the export by reading the same CSV that has been exported.

In [42]:
pd.read_csv('../Resources/Exports/batsman_runs_history.csv').head()

Unnamed: 0         batsman  batsman_runs
0           0  A Ashish Reddy           280
1           1      A Chandila             4
2           2        A Chopra            53
3           3     A Choudhary            25
4           4     A Dananjaya             4

Now, there is a clear problem we can observe in the imported dataframe. It is containing a column named `Unnamed: 0`, that is just containing the index. So, to overcome this, we can use the `index=False` parameter, which implies that index of the DataFrame will not be exported while exporting CSV.

In [43]:
batsman_runs_history.to_csv('../Resources/Exports/batsman_runs_history.csv',
                            index=False)

Now, lets verify it again.

In [44]:
pd.read_csv('../Resources/Exports/batsman_runs_history.csv').head()

batsman  batsman_runs
0  A Ashish Reddy           280
1      A Chandila             4
2        A Chopra            53
3     A Choudhary            25
4     A Dananjaya             4

#### **2. The `columns` parameter**

Sometimes, we might need to export selective columns from the DataFrame to CSV. So, we do have a parameter for this problem.  
The `columns` parameter helps us to export selected columns from the DataFrame as CSV.

Here is a demonstration.

In [45]:
columns = ['wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs',
           'penalty_runs', 'batsman_runs', 'extra_runs', 'total_runs']
runs_history = del_df.groupby('batsman')[columns].sum().reset_index()
runs_history.head()

batsman  wide_runs  bye_runs  legbye_runs  noball_runs  \
0  A Ashish Reddy          3         5            0            0   
1      A Chandila          0         0            0            0   
2        A Chopra          4         0            1            0   
3     A Choudhary          0         0            0            0   
4     A Dananjaya          0         0            0            0   

   penalty_runs  batsman_runs  extra_runs  total_runs  
0             0           280           8         288  
1             0             4           0           4  
2             0            53           5          58  
3             0            25           0          25  
4             0             4           0           4

Now, lets say out of the above DataFrame, I just want to export `wide_runs` history of batsmen. So, we can use the `columns` parameter to export only the `wide_runs` columns.

In [46]:
runs_history.to_csv('../Resources/Exports/wide_runs_histroy.csv',
                    index=False,
                    columns=['batsman', 'wide_runs'])

Lets verify the export.

In [47]:
pd.read_csv('../Resources/Exports/wide_runs_histroy.csv').head()

batsman  wide_runs
0  A Ashish Reddy          3
1      A Chandila          0
2        A Chopra          4
3     A Choudhary          0
4     A Dananjaya          0

#### **3. The `header` parameter**

This is the parameter that writes the column names to the CSV.  
By default, its values is set to `True`, so it uses the column names from the DataFrame to write in the CSV.  
If we set it to `False`, it won't write the column names to the CSV.  
If we provide it an exclusive list of strings, it will treat it as aliases for the column names.

Lets use the `runs_history` for the `header` parameter demonstration.

In [48]:
columns = ['wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs',
           'penalty_runs', 'batsman_runs', 'extra_runs', 'total_runs']
runs_history = del_df.groupby('batsman')[columns].sum().reset_index()
runs_history.head()

batsman  wide_runs  bye_runs  legbye_runs  noball_runs  \
0  A Ashish Reddy          3         5            0            0   
1      A Chandila          0         0            0            0   
2        A Chopra          4         0            1            0   
3     A Choudhary          0         0            0            0   
4     A Dananjaya          0         0            0            0   

   penalty_runs  batsman_runs  extra_runs  total_runs  
0             0           280           8         288  
1             0             4           0           4  
2             0            53           5          58  
3             0            25           0          25  
4             0             4           0           4

Lets, use the default behaviour of `header` parameter.

In [49]:
runs_history.to_csv('../Resources/Exports/runs_history_with_header.csv',
                    index=False)

Lets verify it.

In [50]:
pd.read_csv('../Resources/Exports/runs_history_with_header.csv').head()

batsman  wide_runs  bye_runs  legbye_runs  noball_runs  \
0  A Ashish Reddy          3         5            0            0   
1      A Chandila          0         0            0            0   
2        A Chopra          4         0            1            0   
3     A Choudhary          0         0            0            0   
4     A Dananjaya          0         0            0            0   

   penalty_runs  batsman_runs  extra_runs  total_runs  
0             0           280           8         288  
1             0             4           0           4  
2             0            53           5          58  
3             0            25           0          25  
4             0             4           0           4

Now, lets use the parameter `header` with boolean value `False`.

In [51]:
runs_history.to_csv('../Resources/Exports/runs_history_without_header.csv',
                    index=False, header=False)

Lets verify it keeping in mind that, there is no header in the CSV.

In [52]:
pd.read_csv('../Resources/Exports/runs_history_without_header.csv',
            header=None).head()

0  1  2  3  4  5    6  7    8
0  A Ashish Reddy  3  5  0  0  0  280  8  288
1      A Chandila  0  0  0  0  0    4  0    4
2        A Chopra  4  0  1  0  0   53  5   58
3     A Choudhary  0  0  0  0  0   25  0   25
4     A Dananjaya  0  0  0  0  0    4  0    4

Now, lets try `header` with a custom list of strings (aliases for the column names).

In [53]:
aliases = ['batsman_name', 'wides', 'byes', 'legbyes', 'noballs',
           'penalties', 'batsman_runs', 'extras', 'total']
runs_history.to_csv('../Resources/Exports/runs_history_with_custom_header.csv',
                    index=False, header=aliases)

Now, lets verify it.

In [54]:
pd.read_csv('../Resources/Exports/runs_history_with_custom_header.csv').head()

batsman_name  wides  byes  legbyes  noballs  penalties  batsman_runs  \
0  A Ashish Reddy      3     5        0        0          0           280   
1      A Chandila      0     0        0        0          0             4   
2        A Chopra      4     0        1        0          0            53   
3     A Choudhary      0     0        0        0          0            25   
4     A Dananjaya      0     0        0        0          0             4   

   extras  total  
0       8    288  
1       0      4  
2       5     58  
3       0     25  
4       0      4

You can see that all the column names are according to the `aliases`, not the default ones, as in DataFrames.

#### **4. The `mode` parameter**

The `mode` parameter decides in what state the file is gonna open.  
1. If the `mode` is set to `w`, which is the write mode, Pandas overrides the data of the file (*in case the file already exists*), or creates a new file and write the data to it (*in case the file does not exists*).
2. If the `mode` is set to `a`, which is the append mode, Pandas appends the data to the end of the file.

To demostrate this, we will be using the `batsman_runs_history` dataframe.

In [55]:
batsman_runs_history = del_df.groupby('batsman')['batsman_runs'].sum().reset_index()
batsman_runs_history.head()

batsman  batsman_runs
0  A Ashish Reddy           280
1      A Chandila             4
2        A Chopra            53
3     A Choudhary            25
4     A Dananjaya             4

In [56]:
batsman_runs_history.shape

(516, 2)

As you can see that in `batsman_runs_history`, there are 516 batsman. Though we can export it all in a single CSV, we are gonna export it in four parts, all of which will go into the same file.

In [57]:
row_count = batsman_runs_history.shape[0]
batch_count = 4
batch_size = row_count // batch_count

for batch in range(batch_count):
    if batch == 0:
        batsman_runs_history.iloc[:batch_size * (batch + 1)].to_csv(
            '../Resources/Exports/batsman_runs_history_mode_param.csv',
            index=False, mode='w')
    else:
        batsman_runs_history.iloc[batch_size * batch: batch_size * (batch + 1)].to_csv(
            '../Resources/Exports/batsman_runs_history_mode_param.csv',
            index=False, mode='a', header=False)

Lets verify it.

In [58]:
pd.read_csv('../Resources/Exports/batsman_runs_history_mode_param.csv').shape

(516, 2)

### 🎯 **Exporting XLSX**

Now, the dataframes that we create during preprocessing or during targeted analysis, can also be saved in an XLSX format.  
To export objects of Pandas classes like DataFrame or Series, we use the `to_excel()` method.  

There are two main parameters of this method, that are gonna be used while exporting.  
1. `excel_writer` parameter : It takes either a file path (*in case of single object to be written*), or an `ExcelWriter` object (*in case of multiple objects to be written*).
2. `sheet_name` parameter : It takes the name of the sheet of the excel file, where the object needs to be written.

#### **1. Exporting Single DataFrame**

To export a single dataframe, you just have to provide the file path to the `excel_writer` object.  
Lets export the `batsman_runs_history` to an excel sheet.

In [59]:
batsman_runs_history = del_df.groupby('batsman')['batsman_runs'].sum().reset_index()
batsman_runs_history.head()

batsman  batsman_runs
0  A Ashish Reddy           280
1      A Chandila             4
2        A Chopra            53
3     A Choudhary            25
4     A Dananjaya             4

In [60]:
batsman_runs_history.to_excel(excel_writer='../Resources/Exports/batsman_runs_history.xlsx',
                              sheet_name='batsman_runs')

The above code will now create an excel file named `batsman_runs_history.xlsx`, and inside it, will create a sheet named `batsman_runs` and dump the data into it. If the `sheet_name` parameter would not be used, then, by default a sheet name `Sheet1` could have created inside the excel file.

#### **2. Exporting Multiple DataFrames**

For this portion, we will export two DataFrames, `batsman_runs_history` and `batsman_runs_against_teams`.

In [61]:
batsman_runs_history = del_df.groupby('batsman')['batsman_runs'].sum().reset_index()
batsman_runs_history.head()

batsman  batsman_runs
0  A Ashish Reddy           280
1      A Chandila             4
2        A Chopra            53
3     A Choudhary            25
4     A Dananjaya             4

In [62]:
batsman_runs_against_teams = del_df.pivot_table(index='batsman', columns='bowling_team',
                                                values='batsman_runs', aggfunc='sum')

Now, lets export the above two dataframes. For this, we will have to use the object of `ExcelWriter` and also the `sheet_name` parameter of the `to_excel()` method.

In [63]:
with pd.ExcelWriter('../Resources/Exports/multiple_dataframes.xlsx') as writer:
    batsman_runs_history.to_excel(
        excel_writer=writer, sheet_name='batsman_runs')
    batsman_runs_against_teams.to_excel(
        excel_writer=writer, sheet_name='batsman_runs_against_teams')

Here is the verification.  

First Sheet
![alt text](<Screenshot 2026-01-14 171614.png>)  

Second Sheet
![alt text](<Screenshot 2026-01-14 171632.png>)

### 🎯 **Exporting HTML**

Sometimes, we might need to present our analyzed data on a website. A website is built using HTML, so we will have to export oyr data in the HTML format. For this purpose, we use `to_html()` method of Pandas DataFrame.

Lets prepare a pivot table containing count of sixes, in each over, against each ball.

In [64]:
sixes_pivot = del_df.query('batsman_runs == 6').pivot_table(index='over', columns='ball',
                                                            values='batsman_runs', aggfunc='count')
sixes_pivot

ball     1      2      3      4      5      6     7    8    9
over                                                         
1      7.0   12.0   27.0   31.0   24.0   20.0  12.0  1.0  NaN
2     26.0   30.0   35.0   43.0   45.0   42.0  10.0  5.0  NaN
3     63.0   46.0   57.0   52.0   48.0   58.0   8.0  2.0  1.0
4     49.0   61.0   51.0   81.0   54.0   53.0  11.0  1.0  NaN
5     54.0   56.0   82.0   64.0   62.0   60.0  10.0  2.0  NaN
6     61.0   82.0   44.0   65.0   56.0   66.0  11.0  1.0  NaN
7     27.0   45.0   34.0   44.0   51.0   28.0   3.0  3.0  NaN
8     44.0   47.0   55.0   49.0   53.0   39.0   7.0  NaN  NaN
9     70.0   56.0   56.0   58.0   52.0   35.0  11.0  NaN  1.0
10    43.0   38.0   60.0   43.0   52.0   54.0   9.0  1.0  NaN
11    59.0   70.0   62.0   54.0   61.0   56.0   8.0  NaN  NaN
12    70.0   58.0   72.0   63.0   65.0   52.0  11.0  1.0  NaN
13    67.0   87.0   58.0   80.0   75.0   65.0  10.0  2.0  NaN
14    83.0   85.0   79.0   74.0   65.0   71.0   7.0  3.0  NaN
15    68.0   93.0   83.0   74.0   64.0   86.0  11.0  1.0  NaN
16    77.0   96.0   95.0   91.0   91.0   78.0  15.0  2.0  NaN
17    84.0   88.0   86.0   87.0  103.0   78.0  20.0  3.0  NaN
18    91.0  111.0  112.0  115.0   94.0  101.0  17.0  1.0  1.0
19    82.0   90.0  114.0  103.0  111.0   98.0  31.0  6.0  3.0
20    99.0   95.0  106.0  113.0   88.0  102.0  24.0  5.0  NaN

Now, lets export it as HTML file.

In [65]:
sixes_pivot.to_html('../Resources/Exports/sixes_in_over_balls.html',
                    classes='clean-table')

After applying a bit of CSS to the exported HTML file, we get to see this.  

![alt text](<Screenshot 2026-01-14 180350.png>)

### 🎯 **Exporting JSON**

As we know that JSON is the universal data format, that can be interpreted and used by all languages for sharing data. So, sometimes we might need to export our data into JSON format. For this purpose, `to_json()` method is used.

In [66]:
batsman_runs_for_team = del_df.groupby(['batting_team','batsman'])['batsman_runs'].sum().unstack()

In [67]:
batsman_runs_for_team.to_json('../Resources/Exports/batsman_runs_for_team.json')

### 🎯 **Exporting SQL**

Just like, we import data from `.sql` files using a SQL based database. We can also export our data to any database and the data will be saved there.

Now, to be clear with the process. We will need the following things to export data to a database in the form of tables.
1. All the setup, that we did in the importing SQL files section above needs to be done so that we have the following things ready with us:
   1. A proper DBMS setup (*in my case - MySQL Server and Workbench*)
   2. Username : (*In my case, its **root***)
   3. Host : (*In my case, its **localhost***)
   4. Port : (*In my case, its **3306***)
   5. Password : (*In my case, its **123456***)
   6. `sqlalchemy` and `mysql-connector-python` installed in your working environment.
2. Now, open MySQL Workbench. Create a database in which you want to save you exported tables. (*In my case, I would be creating a database named `ipl`*)  
   ![alt text](<../Resources/Images/Screenshot 2026-01-16 005611.png>)
3. Now, follow the following steps.

4. Create an engine/connection with the help of `create_engine()` method of `sqlalchemy` library.

In [68]:
from sqlalchemy import create_engine
engine = create_engine('mysql+mysqlconnector://root:62292003@localhost:3306/ipl')

5. Now use the `to_sql()` method to export your Pandas DataFrame to SQL Tables.  
   The `to_sql()` method allows us to export our DataFrames to Databases in the form of SQL Tables.  
   It takes the following arguments to function properly:  
   1. `name` parameter wants to know about the name of the table in which you want to save your DataFrame. The `name` you provide may exists or not exists as a table in the database.
   2. `con` parameter requires an SQLAlchemy engine or connection, that has to be connected to the database where the table you provided exists.
   3. `if_exists` parameter is used in the case, where the table with provided `name` already exists. It provided us three ways to deal with this kind of situation:
      1. If it is set to `fail` (Default), it is gonna throw an error, showing that the table already exists.
      2. If it is set to `replace`, it is gonna drop the existing table and dump the new data.
      3. If it is set to `append`, it is gonna insert new values to the existing table. (Old data will be preserved)

Now, lets export some of the DataFrames to MySQL Database.

Lets export the original deliveries, that is, `del_df` dataset to MySQL Database.

In [69]:
del_df.to_sql(name='deliveries', con=engine,
              index=False, chunksize=1000,
              if_exists='replace')

-180

The reason I used the `chunksize` parameter, is that there is a huge no. of rows in the DataFrame, that exceeds the default insertion of rows at a time. So, I set the `chunksize` parameter to `1000`, such that only `1000` rows are inserted at a time in the database.  

Here is the verification of the Export.  
![alt text](<../Resources/Images/Screenshot 2026-01-16 011604.png>)

In [70]:
del_df.shape

(179078, 21)

The row count in MySQL Database matches the row count of `del_df`.

Lets export the `runs_history` dataset to MySQL. 

In [71]:
columns = ['wide_runs', 'bye_runs', 'legbye_runs', 'noball_runs',
           'penalty_runs', 'batsman_runs', 'extra_runs', 'total_runs']
runs_history = del_df.groupby('batsman')[columns].sum().reset_index()

In [72]:
runs_history.to_sql(name='runs_history', con=engine,
                    index=False, if_exists='replace')

-1

Here is the verification for the `runs_history` table.  
![alt text](<../Resources/Images/Screenshot 2026-01-16 012432.png>)

In [73]:
runs_history.shape

(516, 9)

The row count in MySQL Database matches the row count of `runs_history`.

## 🧶 **Fetching Data with API**

In [ ]:
import os
import time
import requests
import pandas as pd
from tqdm import tqdm


def fetch_movie_data():
    BASE_URL = "https://api.themoviedb.org/3/movie/top_rated"
    params = {
        "language": "en-US",
        "page": 1
    }
    headers = {
        "accept": "application/json",
        "Authorization": "Bearer API_ACCESS_TOKEN"
    }
    
    # Initial request to get total pages
    response = requests.get(url=BASE_URL, params=params, headers=headers)
    
    if response.status_code != 200:
        raise Exception(f"API request failed with status {response.status_code}")
    
    page_count = response.json()['total_pages']
    columns = ['id', 'original_language', 'original_title', 'overview',
               'popularity', 'release_date', 'vote_average', 'vote_count']
    
    output_path = '../Resources/Exports/movies_df.csv'
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    
    # Track if file exists to determine write mode
    file_exists = os.path.exists(output_path)
    if file_exists:
        os.remove(output_path)  # Start fresh
    
    for page in tqdm(range(1, page_count + 1), desc="Fetching movies"):
        try:
            params['page'] = page
            page_response = requests.get(
                url=BASE_URL, params=params, headers=headers)
            page_response.raise_for_status()
            
            page_df = pd.DataFrame(
                data=page_response.json()['results'], 
                columns=columns
            )
            
            # Write to CSV incrementally
            page_df.to_csv(
                output_path,
                mode='a',
                header=(page == 1),
                index=False
            )
            
            time.sleep(1)
            
        except requests.exceptions.RequestException as e:
            print(f"\\nNetwork error on Page {page}: {e}")
        except KeyError as e:
            print(f"\\nData parsing error on Page {page}: {e}")
        except Exception as e:
            print(f'\\nUnexpected error on Page {page}: {e}')
    
    # Read and return the complete dataframe
    movies_df = pd.read_csv(output_path)
    print(f"\\nTotal movies fetched: {len(movies_df)}")
    return movies_df

In [ ]:
fetched_df = fetch_movie_data()
fetched_df.shape

## 🧶 **Web Scraping**

Web Scraping is a technique that is used to scrape data from a webpage.

For this section, we will be using `BeautifulSoup` to scrape out data of companies from a website known as Ambition Box. Ambition Box has a page on their website listing companies along with some information about the company.  
Here is the link to the companies page : [Companies - Ambition Box](https://www.ambitionbox.com/list-of-companies?page=1)

Lets write the logic to scrape the data.

In [205]:
import os
import time
import tqdm
import requests
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup


def scrape_companies_details(url: str,
                             params: dict,
                             headers: dict,
                             pages: int):
    '''Scrape Company Details'''
    
    # Ensure Directory
    output_path = '../Resources/Exports/companies.csv'
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    
    # Remove Directory for Fresh Start
    if os.path.exists(output_path):
        os.remove(output_path)

    with tqdm.tqdm(total=100, desc='Fetching Companies') as pbar:
        for page in range(1, pages + 1):
            
            # Get HTML For Page
            params['page'] = page
            webpage = requests.get(url=url,
                                   params=params,
                                   headers=headers).text

            # Making Soup
            soup = BeautifulSoup(markup=webpage, features='html.parser')

            company_cards = soup.find_all(
                name='div', class_='companyCardWrapper')

            company_details = dict(
                company_names=list(),
                company_ratings=list(),
                company_rating_count=list(),
                company_type=list(),
                company_locations=list(),
                average_salary=list(),
                jobs_count=list()
            )
            
            # Process Company Cards
            for company in company_cards:
                
                # Scrape Company Details
                try:
                    company_details['company_names'].append(
                        company.find(name='h2',
                                    class_='companyCardWrapper__companyName').text.strip()
                    )
                except:
                    company_details['company_names'].append(np.nan)
                try:
                    company_details['company_ratings'].append(
                        company.find(name='div',
                                    class_='rating_text').text.strip()
                    )
                except:
                    company_details['company_ratings'].append(np.nan)
                try:
                    company_details['company_rating_count'].append(
                        company.find(name='span',
                                    class_='companyCardWrapper__companyRatingCount').text.strip()[1:-1]
                    )
                except:
                    company_details['company_rating_count'].append(np.nan)
                try:
                    company_details['company_type'].append(
                        company.find(name='span',
                                    class_='companyCardWrapper__interLinking').text.split('|')[0].strip()
                    )
                except:
                    company_details['company_type'].append(np.nan)
                try:
                    company_details['company_locations'].append(
                        company.find(name='span',
                                    class_='companyCardWrapper__interLinking').text.split('|')[1].strip()
                    )
                except:
                    company_details['company_locations'].append(np.nan)
                try:
                    company_details['average_salary'].append(
                        company.find_all(name='span',
                                        class_='companyCardWrapper__ActionCount')[1].text.strip()
                    )
                except:
                    company_details['average_salary'].append(np.nan)
                try:
                    company_details['jobs_count'].append(
                        company.find_all(name='span',
                                        class_='companyCardWrapper__ActionCount')[3].text.strip()
                    )
                except:
                    company_details['jobs_count'].append(np.nan)
                
            # Write to CSV Incrementally
            page_df = pd.DataFrame(company_details)
            page_df.to_csv(
                output_path,
                mode='a',
                index=False,
                header=(page == 1)
            )
            
            # Delay
            time.sleep(0.2)
            
            # Update Progress Bar
            pbar.update(100 / pages)
            
    # Read & Return Data
    companies_df = pd.read_csv(output_path)
    return companies_df

Now, lets call the `scrape_companies_details` method for scraping companies details from `500` pages.

In [ ]:
URL = 'https://www.ambitionbox.com/list-of-companies'
params = {'page': 1}
headers = {'User-Agent': 'Safari/537.36'}
companies_df = scrape_companies_details(url=URL,
                                        params=params,
                                        headers=headers,
                                        pages=500)
companies_df.shape